Launchpad: Paid newsletter task opens modal

edanzer opened the pull request:
## Proposed Changes

Update the `Create paid newsletter` task on Launchpad to directly open the Plans modal rather than redirect the user to the Earn page in the WordPress.com admin area.

<img width="1476" alt="paid newsletter modal" src="https://github.com/Automattic/wp-calypso/assets/21228350/666b2d04-7cb3-4316-8865-55d18fded791">

## Testing Instructions

1) Setup
   - Add `define( 'USE_STORE_SANDBOX', true );` to your sandbox. You can add it to wp-content/mu-plugins/0-sandbox.php. This will allow you to NOT have to fully set up stripe when sent to stripe.
   - Sandbox public-api.wordpress.com and the domain of your test site.
   - Checkout this calypso branch and run yarn or yarn start if needed.
   - Start a newsletter site at `http://calypso.localhost:3000/setup/newsletter/intro`. Be sure to choose the "Paid" option. And remember to sandbox this domain before clicking the stripe link if you've created a new site.
   - Click the `Set up payment method` link. You'll need this before you can add plans. It will take you to the Stripe connect page. If you've set the USE_STORE_SANDBOX constant, you should see a button near the top to Skip this Form. Click that and return to Launchpad.
   - When Stripe sends you back to WordPress, it will load wordpress.com rather than Calypso localhost for the domain. If needed, replace https://wordpress.com with http://calypso.localhost:3000/. 
   
2) Test the link
   - Click the `Create paid newsletter` link
   - Add a plan/product and click save. confirm the Modal closes and that the task now shows as complete on Launchpad. There may be a small delay on completion status.
   - Go to site's admin, to Tools > Earn, and confirm the new plan has been created there. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

edanzer committed:
Make launchpad paid plans link open modal

edanzer committed:
Move modal, invalidate query

edanzer committed:
Move query invalidation

edanzer committed:
Set newsletter toggle on by default

edanzer committed:
Ensure site id is set for api request

simison commented:
Needs a rebase

simison commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/sidebar.tsx:
```diff
@@ -240,6 +247,13 @@ const Sidebar = ( { sidebarDomain, siteSlug, submit, goToStep, flow }: SidebarPr
+					<RecurringPaymentsPlanAddEditModal
+						closeDialog={ () => setShowPlansModal( false ) }
+						product={ { subscribe_as_site_subscriber: true } }
+						siteId={ site?.ID }
```
The site ID seems critical for the modal and methods like `requestAddProduct` that it's using, so we shouldn't rely on optional chaining here. Instead, can we add the check onto `showPlansModal` or otherwise ensure we have the ID?

edanzer replied:
Add this check!

simison commented:
Small CSS issue — for some reason the _Currency_ input doesn't stay next to _Renewal frequency_ input:

At Earn

<img width="644" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/e60a482b-9bd0-4eee-b59c-5054e8a77127">

At Launchpad
<img width="568" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/83699217-a2df-4018-9b7b-5048660a43cd">

simison reviewed (approved):
Tested, works well. Feel free to merge. 

I left couple notes on CSS and the one about ID which I'd protect against better.

I don't think those are blockers — feel free to address either in this PR before merge, or a follow-up PR.

Do re-test after debase as I see some launchpad related things have changed slightly.

simison commented:
It's a bit weird how one can create a new payment plan by re-clicking the task. It's not so obvious you're creating a second one:

<img width="500" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/88c9a2c6-73af-48ca-8e2f-4b36dc4135d7">

<img width="500" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/7aa3d592-441d-47b4-adfa-22be6f55dc0e">

crisbusquets commented:
> It's a bit weird how one can create a new payment plan by re-clicking the task. It's not so obvious you're creating a second one:

If I remember correctly, all launchpad tasks can be clicked again when they're `done` (cc @nuriapenya). Maybe in this case we shouldn't allow it.

edanzer committed:
Ensure siteId exists, resolve TS error

edanzer committed:
Fix payment modal styling

edanzer committed:
Update checks needed for payment modal

edanzer committed:
Fix function argument order

edanzer committed:
Fix unit tests

edanzer commented:
Updates: 
 - Rebased and fixed merge conflicts
 - Added extra check to ensure we're passing a site ID
 - Fixed CSS issue with fields being stack vs side-by-side
 - Fixed failing Launchpad unit tests

simison commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/sidebar.tsx:
```diff
@@ -247,11 +246,15 @@ const Sidebar = ( { sidebarDomain, siteSlug, submit, goToStep, flow }: SidebarPr
-						product={ { subscribe_as_site_subscriber: true } }
-						siteId={ site?.ID }
+						// This spread syntax for props is needed to
+						// resolve an odd IntrinsicAttributes TS error
```
I don't understand this bit, and it feels like a workaround 🤔 

Maybe Folks at Typescript Slack channel can help?

edanzer replied:
It is definitely a work around. Part of the challenge with a real fix is it would almost certainly require updating the underlying component and how props are passed/defined. However, the underlying component: 
  - Is written in JS, not typescript
  - Is a class component
  - Uses connect() and mapStatetoProps to populate props, particularly siteID (prior to our current usage, I think it was assumed this component would always be used on the admin where we know the site ID). 
  
So there are a lot of odd, complicating issues with resolving it. 

I can do spend the time to resolve this if we like. I'm not sure what all will be involved yet, but it probably adds a fair dose of extra time before merging this.

simison replied:
Would a `@ts-ignore` comment work/look better? 🤔  They are rare in the codebase but we do use them in special situations.

edanzer replied:
That would work. I guess the question is if that's better than what I'm already doing? I don't have strong feelings, so I could go any direction. But the @ts-ignore seems reasonable here given the underlying component is not TS.

simison replied:
I'm not proficient enough in TS yet to answer confidently. :-) 

I think the "ignore" is more the intent here, and meanwhile we would want to keep writing our code like we normally would?

edanzer replied:
Yes, that works for me. If you like, I can proceed with that for now to get this merged, but follow up on this in the TypeScript slack channel. If someone has thoughts on a cleaner approach that doesn't involve a major refactor of the underlying component, then handle that in separate PR.

edanzer replied:
Worth noting that adding ts-ignore generates another lint error: `Do not use "@ts-ignore" because it alters compilation errors.eslint[@typescript-eslint/ban-ts-comment](https://typescript-eslint.io/rules/ban-ts-comment)`

simison replied:
Check other usages of `@ts-ignore` in calypso, they 're generally:
```js
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore Explanation why TS was ignored.
```

edanzer replied:
Yup, that's what I did here.

simison commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/style.scss:
```diff
@@ -484,3 +484,29 @@
+// Fix small CSS with payment modal. Ideally, should
+// be resolved via css for payment modal component itself.
+// But that modal doesn't have its own css. The CSS is part
+// of the Earn > Memberships page.
```
Instead of copying the CSS here, can you move the CSS to a new file and import from `my-sites/earn/memberships/add-edit-plan-modal`? This is fine for merging but would need a follow-up to restructure.

Restructuring will be helpful for future with this modal as well.

edanzer replied:
Yes, I thought about doing that here, but I didn't want to mix the point of this PR with refactoring the payments modal component to have it's own CSS. I can handle in a follow up.

simison replied:
Yup that works! 👍

edanzer committed:
Add ts-ignore in place of work around

edanzer committed:
Add explanation for ts-ignore
